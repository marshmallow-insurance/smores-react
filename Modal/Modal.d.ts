import { FC, ReactNode } from 'react';
export type ModalProps = {
    title?: string;
    icon?: string;
    children?: ReactNode;
    showModal?: boolean;
    handleClick: () => void;
    drawer?: boolean;
    cross?: boolean;
    width?: string;
    containerClass?: string;
    portalContainer?: Element | DocumentFragment;
};
export declare const Modal: FC<ModalProps>;
