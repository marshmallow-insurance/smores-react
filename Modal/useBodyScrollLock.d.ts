export default function useBodyScrollLock({ node, showModal, }: {
    node: HTMLDivElement | null;
    showModal: boolean;
}): void;
