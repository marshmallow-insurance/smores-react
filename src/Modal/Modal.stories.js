import React, { useState } from 'react'
import { Modal } from '../Modal'
import { Button } from '../Button'
import { Box } from '../Box'

export default {
  title: 'Modal'
}

const Template = (args) => {
  const [showModal, setShowModal] = useState(false);

  const handleClick = () => {
    setShowModal(!showModal)
  }
  return (
    <Box>
      {showModal ? (
        <Modal {...args} handleClick={handleClick} showModal={showModal}>
          [A modal window] creates a mode that disables the main window, but keeps it visible with the modal window as a child window in front of it.
        </Modal>
      ) : (
        <Button primary handleClick={handleClick}>Show Modal</Button>
      )}
    </Box>
  )
}

export const Default = Template.bind({})

Default.args = {
  title: "Hello world i'm a beautiful modal",
  showModal: false,
}
