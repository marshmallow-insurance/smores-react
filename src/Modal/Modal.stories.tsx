import React, { useState } from 'react'
import { Button } from '../Button'
import { Box } from '../Box'
import { Modal, ModalProps } from './Modal'

export default {
  title: 'Modal',
  component: Modal,
  argTypes: { handleClick: { action: 'Callback - `handleClick`' } },
}

const Template = (props: ModalProps) => {
  const [showModal, setShowModal] = useState(false)

  const handleClick = () => {
    setShowModal(!showModal)
  }
  return (
    <Box height="900px">
      <Modal {...props} handleClick={handleClick} showModal={showModal}>
        [A modal window] creates a mode that disables the main window, but keeps
        it visible with the modal window as a child window in front of it.
      </Modal>
      <Button primary handleClick={handleClick}>
        Show Modal with Mobile Drawer
      </Button>
    </Box>
  )
}

export const Interactive = Template.bind({})

Interactive.args = {
  icon: 'calendar',
  title: "Hello world i'm a beautiful modal",
  showModal: false,
  drawer: true,
  cross: true,
  width: '640px',
}
