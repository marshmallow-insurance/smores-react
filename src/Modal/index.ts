export { Modal } from './Modal'
export { useBodyScrollLock } from './useBodyScrollLock'
