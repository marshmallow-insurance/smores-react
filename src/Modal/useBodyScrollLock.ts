import { useEffect } from 'react'
import {
  clearAllBodyScrollLocks,
  disableBodyScroll,
  enableBodyScroll,
} from 'body-scroll-lock'

export default function useBodyScrollLock({
  node,
  showModal,
}: {
  node: HTMLDivElement | null
  showModal: boolean
}) {
  useEffect(() => {
    return () => clearAllBodyScrollLocks()
  }, [])

  useEffect(() => {
    if (node === null) return

    if (showModal) {
      disableBodyScroll(node, {
        reserveScrollBarGap: true,
        allowTouchMove: () => true,
      })
    } else {
      enableBodyScroll(node)
    }
  }, [node, showModal])
}
