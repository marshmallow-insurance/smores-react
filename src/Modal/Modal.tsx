import React, { FC, ReactNode, useRef } from 'react'
import { createPortal } from 'react-dom'
import styled, { css } from 'styled-components'

import { Box } from '../Box'
import { IconStrict } from '../IconStrict'
import { Text, type TextProps } from '../Text'
import { useBodyScrollLock } from '../hooks/useBodyScrollLock'

interface IModalContainer {
  // drawer state
  $drawer: boolean
  // modal width
  $width: string
}

export type ModalProps = {
  /**
   * Title of the modal
   * @default "" (empty string)
   *
   * @example
   * ```tsx
   * <Modal title="MultiCar Account" />
   * ```
   *
   * @example
   * ```tsx
   * <Modal title={{ typo: 'hero', children: 'MultiCar Account' }} />
   * ```
   */
  title?: string | TitleProps
  icon?: string
  children?: ReactNode
  rightPanel?: ReactNode
  showModal?: boolean
  handleClick: () => void
  drawer?: boolean
  cross?: boolean
  width?: string
  containerClass?: string
  portalContainer?: Element | DocumentFragment
  closeOnOverlayClick?: boolean
}

export type TitleProps = TextProps

const getDefaultTitleProps = (title: string): TitleProps => ({
  children: title,
  tag: 'h2',
  typo: 'heading-small',
  align: 'left',
})

export const Modal: FC<ModalProps> = ({
  title = '',
  children,
  rightPanel,
  showModal = false,
  handleClick,
  drawer = true,
  cross = true,
  width,
  containerClass,
  portalContainer = document.body,
  closeOnOverlayClick = true,
}) => {
  const modalRef = useRef<HTMLDivElement>(null)

  useBodyScrollLock({ node: modalRef.current, showModal })

  const isTitleString = typeof title === 'string'
  const titleProps = isTitleString ? getDefaultTitleProps(title) : title

  if (!showModal) return null

  return createPortal(
    <Wrapper ref={modalRef}>
      <Overlay
        onClick={() => closeOnOverlayClick && handleClick()}
        $closeOnOverlayClick={closeOnOverlayClick}
      />
      <Container
        $drawer={drawer}
        $width={width || '460px'}
        className={containerClass}
      >
        <Box
          flex
          alignItems="flex-start"
          justifyContent="space-between"
          mb="8px"
        >
          <TitleElements flex direction="column">
            <Text {...titleProps} />
          </TitleElements>
          <Box flex alignItems="center" gap={'8px'}>
            {rightPanel}
            {cross && (
              <IconStrict
                render="cross"
                title="Close modal"
                backgroundColor="oatmeal"
                handleClick={handleClick}
                size={36}
              />
            )}
          </Box>
        </Box>
        <Box flex direction="column">
          {children}
        </Box>
      </Container>
    </Wrapper>,
    portalContainer,
  )
}

const Wrapper = styled(Box)`
  display: flex;
  position: absolute;
  z-index: 999;
  top: 0;
  left: 0;
  height: 100vh;
  width: 100%;
  justify-content: center;
  align-items: center;
`

const Overlay = styled.div<{ $closeOnOverlayClick: boolean }>`
  position: fixed;
  background: ${({ theme }) => theme.color.surface.base[900]};
  cursor: ${(props) => (props.$closeOnOverlayClick ? 'pointer' : 'default')};
  opacity: 0.4;
  top: 0;
  bottom: 0;
  left: 0;
  right: 0;
`

const Container = styled.div<IModalContainer>(
  ({ $drawer, $width }) => css`
    background: ${({ theme }) => theme.color.background[100]};
    box-sizing: border-box;
    border-radius: 16px;
    padding: 24px;
    width: 100%;
    max-width: ${$width};
    position: fixed;
    max-height: calc(100vh - 64px);
    overflow: auto;
    transition: all 0.3s ease-in-out;

    ${$drawer === true &&
    css`
      @media (max-width: 768px) {
        max-width: none;
        border-radius: 16px 16px 0px 0px;
        padding: 10% 24px;
        max-height: 90vh;

        position: fixed;
        right: 0;
        left: 0;
        bottom: 0;
      }
    `}
  `,
)

const TitleElements = styled(Box)`
  align-self: center;
`
