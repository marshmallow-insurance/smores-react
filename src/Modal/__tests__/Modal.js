import React from 'react';
import { render } from '@testing-library/react';
import 'jest-styled-components';

import {Modal} from '../Modal';

test('renders', () => {
  const {container} = render(<Modal>Text inside</Modal>);
  expect(container.firstChild).toMatchSnapshot();
});
