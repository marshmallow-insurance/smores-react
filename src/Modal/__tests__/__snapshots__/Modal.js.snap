// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`renders 1`] = `
.c3 {
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-box-pack: justify;
  -webkit-justify-content: space-between;
  -ms-flex-pack: justify;
  justify-content: space-between;
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
}

.c6 {
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
  -webkit-box-pack: center;
  -webkit-justify-content: center;
  -ms-flex-pack: center;
  justify-content: center;
  width: 100%;
  max-width: 32px;
  height: 32px;
  -webkit-transform: rotate(0deg);
  -ms-transform: rotate(0deg);
  transform: rotate(0deg);
  margin-top: 0;
  margin-right: 0;
  margin-bottom: 0;
  margin-left: 0;
}

.c4 {
  font-size: 18px;
  line-height: 23px;
  font-weight: 500;
  margin: 0;
  padding: 0;
  color: #112035;
  text-align: left;
  cursor: inherit;
  -webkit-letter-spacing: 0.15px;
  -moz-letter-spacing: 0.15px;
  -ms-letter-spacing: 0.15px;
  letter-spacing: 0.15px;
}

.c7 {
  font-size: 12px;
  line-height: 16px;
  font-weight: 400;
  margin: 0;
  padding: 0;
  color: #636768;
  text-align: left;
  cursor: inherit;
  -webkit-letter-spacing: 0.15px;
  -moz-letter-spacing: 0.15px;
  -ms-letter-spacing: 0.15px;
  letter-spacing: 0.15px;
}

.c0 {
  display: none;
  position: absolute;
  height: calc(100vh);
  width: 100%;
}

.c1 {
  position: absolute;
  background: #112035;
  opacity: 0.4;
  height: calc(100vh);
  width: 100%;
}

.c2 {
  background: #ffffff;
  z-index: 999;
  position: absolute;
  border: 1px solid #E5E5E5;
  box-sizing: border-box;
  border-radius: 8px;
  padding: 24px;
  max-width: 456px;
  left: calc(50% - 460px / 2);
  top: calc(50% - 265px / 2 + 0.5px);
}

.c5 {
  cursor: pointer;
  background: #F9F9F9;
  border-radius: 32px;
  margin: 8px;
}

.c5:hover {
  background: #F8F6F6;
}

@media (min-width:768px) {
  .c4 {
    font-size: 21px;
    line-height: 27px;
  }
}

@media (min-width:768px) {
  .c7 {
    font-size: 14px;
    line-height: 18px;
  }
}

@media (max-width:768px) {
  .c2 {
    max-width: none;
    width: 100%;
    left: 0;
    border-radius: 16px 16px 0px 0px;
    padding: 10% 24px;
    overflow: scroll;
    max-height: 90vh;
    height: 70vh;
    top: 30vh;
  }
}

<div
  class="c0"
>
  <div
    class="c1"
  />
  <div
    class="c2"
  >
    <div
      class="c3"
    >
      <h2
        class="c4"
        color="blue7"
        cursor="inherit"
      />
      <div
        class="c5"
      >
        <span
          class="c6"
          color="blue7"
          rotate="0"
          size="32"
        >
          <svg
            fill="none"
            height="100%"
            viewBox="0 0 32 32"
            width="100%"
          >
            <path
              d="M17.4139 15.5L22.7068 9.82913C23.0977 9.41022 23.0977 8.7331 22.7068 8.31418C22.3158 7.89527 21.6838 7.89527 21.2928 8.31418L16 13.985L10.7072 8.31418C10.3162 7.89527 9.68423 7.89527 9.29324 8.31418C8.90225 8.7331 8.90225 9.41022 9.29324 9.82913L14.5861 15.5L9.29324 21.1709C8.90225 21.5898 8.90225 22.2669 9.29324 22.6858C9.48823 22.8947 9.74422 22.9997 10.0002 22.9997C10.2562 22.9997 10.5122 22.8947 10.7072 22.6858L16 17.0149L21.2928 22.6858C21.4878 22.8947 21.7438 22.9997 21.9998 22.9997C22.2558 22.9997 22.5118 22.8947 22.7068 22.6858C23.0977 22.2669 23.0977 21.5898 22.7068 21.1709L17.4139 15.5Z"
              fill="#112035"
            />
          </svg>
        </span>
      </div>
    </div>
    <p
      class="c7"
      color="grey8"
      cursor="inherit"
    >
      Text inside
    </p>
  </div>
</div>
`;
